sn_interface/src/network_knowledge/section_tree/mod.rs

```diff
@@ -167,24 +168,45 @@ impl SectionTree {
     pub fn update(
         &mut self,
         signed_sap: SectionAuth<SectionAuthorityProvider>,
-        proof_chain: &SecuredLinkedList,
+        partial_dag: &SectionsDAG,
     ) -> Result<bool> {
         // Check if SAP signature is valid
         if !signed_sap.self_verify() {
             return Err(Error::UntrustedSectionAuthProvider(format!(
-                "invalid signature: {:?}",
+                "Invalid signature: {:?}",
                 signed_sap.value
             )));
         }
 
         // Check if SAP's section key matches SAP signature's key
         if signed_sap.sig.public_key != signed_sap.section_key() {
             return Err(Error::UntrustedSectionAuthProvider(format!(
-                "section key doesn't match signature's key: {:?}",
+                "Section key doesn't match signature's key: {:?}",
                 signed_sap.value
             )));
         }
 
+        // Make sure the proof chain can be trusted,
+        // i.e. check each key is signed by its parent/predecessor key.
+        if !partial_dag.self_verify() {
+            return Err(Error::UntrustedPartialDAG(format!(
+                "Invalid partial_dag: {:?}",
+                partial_dag
+            )));
+        }
+
+        // Check the SAP's key is the last key of the proof chain
+        if partial_dag.leaf_keys().len() != 1 {
+            return Err(Error::MultipleBranchError);
+        }
+        if !partial_dag.leaf_keys().contains(&signed_sap.section_key()) {
```

Comment:
I think you can combine these last two checks with
```suggestion
        if partial_dag.leaf_keys() != BTreeSet::from_iter([signed_sap.section_key()]) {
```
Also make sure to log `partial_dag.leaf_keys()` in the error msg string as well.

Reply:
Since `partial_dag` can contain a tree with more than one branch, the `MultipleBranchError` was introduced to specify that we expect a dag with just 1 branch. So, is it ok to separate out the errors?

Reply:
ok, yeah makes sense, I saw you testing for that error variant in some test cases.